engines/ml/lightgbm/src/main/java/ai/djl/ml/lightgbm/LgbmNDArray.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.ml.lightgbm;
+
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.NDArrayAdapter;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import ai.djl.ndarray.types.SparseFormat;
+
+import com.microsoft.ml.lightgbm.SWIGTYPE_p_double;
+import com.microsoft.ml.lightgbm.SWIGTYPE_p_float;
+import com.microsoft.ml.lightgbm.SWIGTYPE_p_void;
+import com.microsoft.ml.lightgbm.lightgbmlib;
+import com.microsoft.ml.lightgbm.lightgbmlibConstants;
+
+import java.nio.ByteBuffer;
+import java.nio.DoubleBuffer;
+import java.nio.FloatBuffer;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicReference;
+
+/** {@code LgbmNDArray} is the LightGBM implementation of {@link NDArray}. */
+public class LgbmNDArray extends NDArrayAdapter {
+
+    private ByteBuffer data;
```

Comment:
This part looks strange to me. Why we are trying to keep a ByteBuffer reference instead of storing a native pointer for the data? And I can see you are trying to pass the data to native only during `getHandle()` function. Shoudn't we just create the handle and pass it to the NDArray?

Reply:
We need keep both handler and ByteBuffer.

The problem the handle is write only, cannot read data from it. It can only used for inference.

So we have to have two type of NDArray:
1. input only NDArraty, we create handle and pass to engine for inference
2. output only NDArray, we get bytes from the model and need return a NDArray, we don't have handle in this case.